client/evilwallet/wallets.go

```diff
@@ -73,28 +74,60 @@ func (w *Wallets) GetUnspentOutput(wallet *Wallet) *Output {
 	return wallet.GetUnspentOutput()
 }
 
-func (w *Wallets) FreshWallet() *Wallet {
-	if len(w.faucetWallets) > 0 {
-		return w.wallets[w.faucetWallets[0]]
+// IsFaucetWalletAvailable checks if there is any faucet wallet left.
+func (w *Wallets) IsFaucetWalletAvailable() bool {
+	return len(w.faucetWallets) > 0
+}
+
+// FreshWallet returns the first non-empty wallet from the faucetWallets queue. If current wallet is empty,
+// it is removed and the next enqueued one is returned.
+func (w *Wallets) FreshWallet() (*Wallet, error) {
+	if !w.IsFaucetWalletAvailable() {
+		return nil, errors.New("no faucet wallets available, need to request more funds")
 	}
-	return nil
+	wallet := w.wallets[w.faucetWallets[0]]
+	if wallet.IsEmpty() {
+		w.removeWallet(fresh)
+		if !w.IsFaucetWalletAvailable() {
+			return nil, errors.New("no faucet wallets available, need to request more funds")
+		}
+		// take next wallet
+		wallet = w.wallets[w.faucetWallets[0]]
+		if wallet.IsEmpty() {
+			return nil, errors.New("wallet is empty, need to request more funds")
+		}
+	}
```

Comment:
`IsFaucetWalletAvailable` should be checked before retrieving the wallet. Then the process can be in a loop, return a valid wallet until we get one, or return an error instead.

Reply:
not sure If I understand correctly, but `IsFaucetWalletAvailable` is already checked in line 85 and the error is returned, as the check simply says that there are no wallets at all, so any further action makes no sense. The empty check is checking if all outputs were used, if yes --> take the next wallet ( that should not be empty - line 96 is just in case but should never happen )

Reply:
Ah ok! then maybe having a `GetWallet` function to retrieve one that is not empty would be cleaner?